Clean up text for neighborhood analysis notebook (#1057)
AlexsLemonade/training-modules: spatial/06-neighborhood-analysis.Rmd

Reviewer: ```suggestion
1. Find the k nearest neighboring cells for each cell (`findNearCells()`).
2. Convert the distances to those neighbors into probabilities, so that closer cells contribute more (`scanHoods()`).
3. Group those per-neighbor probabilities by cell type, giving each cell a probability of being surrounded by each cell type (`mergeByGroup()`).
```

Reviewer: wow today we are learning that suggestions on lines you didn't modify come out _wrong_.....

Reviewer: ```suggestion
`hoodscanR` expects the annotation it should use to live in a standardized `colData` column called `cell_annotation`, and it expects the spatial coordinates to be in the `spatialCoords` of the object.
```

Reviewer: Again github has chosen to make this suggestion render unhelpfully?

Reviewer: You probably mean "data frame" here

Reviewer: This is the first time entropy is mentioned in the notebook - I'd add an explanation above where you explain perplexity.
Edit, but since entropy isn't actually used, I'd point that out too that this function will give it to you but we're focusing on perplexity for reasons.

> In this notebook we'll use the [`hoodscanR`](https://www.bioconductor.org/packages/release/bioc/html/hoodscanR.html) package to look at the following characteristics.

Reviewer: ```suggestion
In this notebook we'll use the [`hoodscanR`](https://www.bioconductor.org/packages/release/bioc/html/hoodscanR.html) package to look at the following characteristics:
```

1. Neighborhood profiles, or the mix of cell types among each cells' nearest neighbors.
2. Neighborhood complexity, which defines if a cell is sitting in a "pure" environment dominated by one cell type, or a highly mixed one.
3. Co-localization, which tells us which cell types tend to show up in the same neighborhoods across the whole slide.
4. Domain detection, which uses clustering of neighborhood profiles to identify spatial domains that share a similar neighborhood profile.

> The general idea behind `hoodscanR` is to look at each cell's k nearest neighbors, convert the distances to those neighbors into probabilities, and then summarize those probabilities by cell type.

Reviewer: I think we want this syntax in general

```suggestion
The general idea behind `hoodscanR` is to look at each cell's k-nearest neighbors, convert the distances to those neighbors into probabilities, and then summarize those probabilities by cell type.
```

> The result is one profile per cell which includes a set of probabilities describing how likely each cell type is to be found in that cell's neighborhood.

Reviewer: The word "likely" is a bit loaded since I don't think we're actually talking about likelihood here, hence suggested rephrase
```suggestion
The result is one profile per cell.
Each profile contains the probability that each cell type present in the data is present in the given cell's neighborhood.
```

spe <- readr::read_rds(normalized_rds)
```

As a reminder, let's look at the cell types that are present in this dataset. 
We'll use the `cell_types_lumped` column that we created in the previous notebook for the plot and for the neighborhood analysis.
For the color palette, we'll read in a TSV file with the colors to use for each cell type. 

> Saving the color palette as a TSV file can be a helpful way to use the same color palettes across multiple notebooks and/or scripts.

Reviewer: ```suggestion
For the color palette, we'll read in a TSV file that we have prepared with the colors to use for each cell type. 
In general, saving color palettes as a TSV files can be a helpful way to use the same color palettes across multiple notebooks and/or scripts.
```

```

Setting the cell type column as a factor lets us control the order the cell types appear in legends and plots.

> We use the order of the palette provided in the TSV file.

Reviewer: ```suggestion
Rather than R defaulting to using alphabetical order, we want to specify that the palette order should match the order in the TSV file.
```

  scale_color_manual(values = cell_type_palette)
```

Note that we still have a fair number of "Unlabeled" cells. 

> From the original publication, the Xenium data was projected onto the single-cell RNA-seq data and the 30 nearest neighbors were identified for each cell.

Reviewer: ```suggestion
From the original publication, the Xenium data was projected onto the single-cell RNA-seq data, and the 30 nearest neighbors were identified for each cell.
```

> If at least half of those neighbors shared the same cell type, then that cell type was used as the cell type label. 

Reviewer: ```suggestion
If at least half of those single-cell neighbors shared the same cell type, then that cell type was used as the cell type label for the Xenium cell. 
```

If this criteria was not met, then the cell was annotated as "Unlabeled". 

> Because all "Unlabeled" cells would have the same cell type, they would be grouped together in the neighborhood analysis. 

Reviewer: ```suggestion
Because all "Unlabeled" cells share this label, they would be grouped together in the neighborhood analysis. 
```

Note that distance here is the Euclidean distance between the x and y coordinates on the slide. 

The choice of `k` sets the size of the neighborhood we're describing.
A small `k` describes the tight ring of cells immediately around each cell, while a large `k` describes a broader region of tissue.

Here we'll use `k = 100`, which is the default for `findNearCells()`.

> One hundred neighbors is large enough that each cell's profile is stable, meaning one unusual neighbor can't dominate it, but still small enough to describe a local region rather than the whole slide.

Reviewer: ```suggestion
100 neighbors is large enough that each cell's profile is stable, meaning one unusual neighbor can't dominate it, but still small enough to describe a local region rather than the whole slide.
```

> When you apply this to your own data, it's worth trying a couple of values of `k` and checking whether the patterns you see hold up.

Reviewer: This is a good point but I would clarify it further. For example, if I were to use 2-3 neighbors only I would probably expect differences from 100 neighbors. So, I think it's not only about whether things hold up but about understanding that the scale can have an effect itself.

neighbors$distance[1:10, 1:5]
```

> Two things are worth noticing about the row and column names here.

Reviewer: Can we make these two things into a numbered list?

The row names are the cell IDs carried over from the SPE object.

> Because we filtered out low quality cells in the previous notebook and removed the "Unlabeled" cells, some IDs may be missing.

Reviewer: ```suggestion
Because we filtered out low-quality cells in the previous notebook and removed the "Unlabeled" cells, some IDs may be missing.
```

> The gaps in ID numbers is expected.

Reviewer: ```suggestion
The jumps in ID numbers are expected.
```

The column names are ranks, not cell IDs.
`nearest_cell_1` is each cell's closest neighbor, `nearest_cell_2` is its second closest, and so on out to the 100th.
This is why the distances increase from left to right in every row.

Next we convert those distances into probabilities with `scanHoods()`.
For each cell, the distances to its neighbors are passed through a softmax transformation, which transforms a vector into a probability distribution where all values sum to 1.
The effect is that nearer cells get more weight than farther ones.

The softmax works in two steps.

* First, each neighbor's distance is turned into a weight that shrinks exponentially as the distance grows, so a cell twice as far away contributes much less than half as much.
* Second, all 100 weights for a cell are divided by their sum, so they add up to 1 and can be read as probabilities.

> The result is that the closest neighbors take up most of the probability and distant neighbors contribute very little.

Reviewer: I would make the point explicitly here that the interplay between k and tau is an important thing to keep tabs on, based on your scientific question.

```

Right now the columns are individual neighbors, which isn't very interpretable on its own.

> `mergeByGroup()` fixes this by adding up the probability weights of all neighbors that share a cell type.

Reviewer: ```suggestion
To get an object we can work with more easily, we'll use the `mergeByGroup()` function which adds up the probability weights of all neighbors that share a cell type.
```

> It needs both the probability matrix, for the weights, and the `cells` matrix, for the cell type of each neighbor.

Reviewer: The commas make this look like a list, but really they are descriptors so I'm suggesting ().

```suggestion
It needs both the probability matrix (for the weights) and the `cells` matrix (for the cell type of each neighbor).
```

```

A nice way to get a feel for these profiles is to plot them as a heatmap, with cells as rows and cell types as columns.

> We'll use the `ComplexHeatmap` package to do this since it has some nice customization options. 

Reviewer: I actually like a lot of this text you removed! It's a more thorough treatment of _why_ we are using complexheatmap. I don't think you need to say it's a wrapper, just that we have more control with complexheatmap directly.

You can learn more about this package, including all of the many plot modifications, from the [`ComplexHeatmap` documentation](https://jokergoo.github.io/ComplexHeatmap-reference/book/). 

Alternatively, `hoodscanR` provides `plotHoodMat()`, which makes a similar plot with fewer options for customization.

We will plot the first 500 cells here just to keep the heatmap legible.
Each column will be a cell and each row will represent a cell type. 
The color indicates the probability of a given cell type appearing in the nearest 100 neighbors of each cell.

```{r neighborhood-heatmap}
# plot the neighborhood profiles for the first 500 cells
ComplexHeatmap::Heatmap(
  t(hoods[1:500,]), # transpose so cells are columns

>   col            = scico::scico(200, palette = "acton", direction = -1),

Reviewer: NB, this seems better than what we have for jaccard in scpca-nf since the low value it's just plain white. A thing to consider.

  name           = "Probability",
  column_title   = "Probability of cell types\n within 100 nearest neighbors",
  column_title_gp = grid::gpar(fontsize = 15),
  column_dend_height = grid::unit(1, "cm"),
  row_dend_width = grid::unit(1, "cm"),

>   column_names_rot = 45,

Reviewer: not used
```suggestion
```

  cluster_rows   = TRUE,     # group cells with similar profiles together
  cluster_columns = TRUE,    # group cell types that co-occur together
  show_column_names = FALSE     # hide per-cell labels; there are too many to read
)
```

This heatmap helps us visualize neighborhood profiles and makes it easy to see when certain cell types might be found in the same neighborhood. 

> For example, there appears to be immune cell neighborhoods where macrophages, T cells, and B cells are all found in the neighborhood of the same cell. 

Reviewer: ```suggestion
For example, there appear to be immune cell neighborhoods where macrophages, T cells, and B cells are all found in the neighborhood of the same cell. 
```

Reviewer: I think there is some insight missing here. We have columns for each cell and can see what surrounds each cell, but it's hard to make a connection from that observation to a wholesale _neighborhood_ without connecting the dots.

Author: I think I disagree with this statement slightly. The heatmap shows the neighborhood of each cell since by definition, we get information about cell types that are present in each cell's local neighborhood, which is what this heatmap is showing. The word neighborhood can both refer to an individual cell's neighborhood or the neighborhood present in a region of the tissue, which is I think what you mean by wholesale neighborhood. I'm going to add a sentence just clarifying that this tells us about individual cells and not cell types as a whole, but I don't think there's much more to add.

```

A nice way to get a feel for these profiles is to plot them as a heatmap, with cells as rows and cell types as columns.
We'll use the `ComplexHeatmap` package to do this since it has some nice customization options, allowing us to have more control over the plot's aesthetics.
You can learn more about this package, including all of the many plot modifications, from the [`ComplexHeatmap` documentation](https://jokergoo.github.io/ComplexHeatmap-reference/book/). 

> Alternatively, `hoodscanR` provides `plotHoodMat()`, which makes a similar plot with fewer options for customization.

Reviewer: The thing that feels off to me here is that "alternatively" is separated onto its own line and makes it seem like a thing we will literally alternatively do. Can we perhaps put it in parentheses the way you do for other hoodscanr functions we don't use but mention? Thinking of this:

```
(`hoodscanR` also provides a function, `hoodscanR::plotTissue()`, to make these plots, but the output is the same).
```

> Alternatively, `hoodscanR` provides `plotHoodMat()`, which makes a similar plot with fewer options for customization.

Author: ```suggestion
You can learn more about this package, including all of the many plot modifications, from the [`ComplexHeatmap` documentation](https://jokergoo.github.io/ComplexHeatmap-reference/book/). 
(`hoodscanR` also provides a function, `hoodscanR::plotTissue()`, to make a similar plot with fewer options for customization).
```

Here we are clustering cells by their neighborhood composition and do not consider gene expression.

As with `k` in the nearest-neighbor step, the number of clusters `k` here is a choice we have to make, and it strongly shapes the result.
To choose `k`, we will cluster the neighborhood profiles using a range of `k` values and create an elbow plot. 

For each value of `k`, k-means assigns every cell to a domain, and we record the total within-cluster sum of squares (WCSS).
The WCSS is the summed squared distance from each cell to the center of the domain it was assigned to, so a smaller value means cells sit closer to their domain's center.

> WCSS always decreases as `k` increases, because more domains means every cell has a nearer center to belong to.

Reviewer: ```suggestion
WCSS always decreases as `k` increases, because more domains means every cell has a closer center to belong to.
```

  labs(fill = "Cell types")
```

This plot is helpful in showing us which cell types contribute to which domain, and ultimately which cell types tend to be found in the same neighborhood profile. 

> Specifically, we see that domain 2 is dominated by tumor cells while domain 4 is a mix of DCIS and Myoepithelial cells. 

Reviewer: ```suggestion
Specifically, we see that domain 2 is dominated by tumor cells, while domain 4 is a mix of DCIS and Myoepithelial cells. 
```

  column_title   = "Probability of cell types\n within 100 nearest neighbors",
  column_title_gp = grid::gpar(fontsize = 15),
  column_dend_height = grid::unit(1, "cm"),
  row_dend_width = grid::unit(1, "cm"),

>   column_names_rot = 45,

Reviewer: this isn't used

```suggestion
```
